### PR TITLE
Add Trinary op template in mshadow

### PR DESCRIPTION
**Pull request by Lorrainexun**

### CODE REVIEW

**Comment by sxjscience** [comment]
How about changing `Trinary` to `Ternary`?

File: mshadow/expr_engine-inl.h [hunk 1]
```diff
@@ -369,7 +407,24 @@ struct ShapeCheck<dim, BinaryMapExp<OP, TA, TB, DType, etype> > {
...
+template<int dim, typename OP, typename TA, typename TB, typename TC,
+         typename DType, int etype>
+struct ShapeCheck<dim, TernaryMapExp<OP, TA, TB, TC, DType, etype> > {
+  inline static Shape<dim>
+  Check(const TernaryMapExp<OP, TA, TB, TC, DType, etype> &t) {
+    Shape<dim> shape1 = ShapeCheck<dim, TA>::Check(t.item1_);
+    Shape<dim> shape2 = ShapeCheck<dim, TB>::Check(t.item2_);
+    Shape<dim> shape3 = ShapeCheck<dim, TC>::Check(t.item3_);
+    bool same = (shape1 == shape2) && (shape2 == shape3);
+    CHECK(same) << "BinaryMapExp: Shapes of operands are not the same, " <<
```
Comment (winstywang) [inline mshadow/expr_engine-inl.h]: Should be Ternary. And remember to output shape3

File: mshadow/expr_engine-inl.h [hunk 2]
```diff
@@ -104,6 +104,22 @@ class Plan<TypecastExp<DstDType, SrcDType, EType, etype>, DstDType> {
...
 };
+
+// ternary expression
+template<typename OP, typename TA, typename TB,typename TC, int etype, typename DType>
+class Plan<TernaryMapExp<OP, TA, TB,TC, DType, etype>, DType> {
+ public:
+  explicit Plan(const Plan<TA, DType> &item1, const Plan<TB, DType> &item2,const Plan<TC, DType> &item3)
+      : item1_(item1), item2_(item2),item3_(item3) {}
+  MSHADOW_XINLINE DType Eval(index_t y, index_t x) const {
+    return OP::Map(item1_.Eval(y, x), item2_.Eval(y, x),item3_.Eval(y,x));
```
Comment (winstywang) [inline mshadow/expr_engine-inl.h]: Space after ,

**Review by winstywang — CHANGES_REQUESTED** [review]
Two minor issues to be fixed. All other parts LGTM.

**Review by winstywang — APPROVED** [review]

**Pull request merged**